src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -691,104 +705,52 @@ InstanceKlass* SystemDictionary::resolve_instance_class_or_null(Symbol* name,
+          // Add the LOAD_INSTANCE token. Threads will wait on loading to complete for this thread.
           PlaceholderEntry* newprobe = placeholders()->find_and_add(name_hash, name, loader_data,
                                                                     PlaceholderTable::LOAD_INSTANCE,
                                                                     NULL,
                                                                     THREAD);
-          load_instance_added = true;
+          load_placeholder_added = true;
         }
       }
     }
 
-    // must throw error outside of owning lock
+    // Must throw error outside of owning lock
     if (throw_circularity_error) {
-      assert(!HAS_PENDING_EXCEPTION && !load_instance_added, "circularity error cleanup");
+      assert(!HAS_PENDING_EXCEPTION && !load_placeholder_added, "circularity error cleanup");
       ResourceMark rm(THREAD);
       THROW_MSG_NULL(vmSymbols::java_lang_ClassCircularityError(), name->as_C_string());
     }
 
-    if (!class_has_been_loaded) {
+    // Be careful when modifying this code: once you have run
+    // placeholders()->find_and_add(PlaceholderTable::LOAD_INSTANCE),
+    // you need to find_and_remove it before returning.
+    // So be careful to not exit with a CHECK_ macro between these calls.
 
+    if (loaded_class == NULL) {
       // Do actual loading
-      loaded_class = load_instance_class(name, class_loader, THREAD);
-
-      // If everything was OK (no exceptions, no null return value), and
-      // class_loader is NOT the defining loader, do a little more bookkeeping.
-      if (!HAS_PENDING_EXCEPTION && loaded_class != NULL &&
-        loaded_class->class_loader() != class_loader()) {
-
-        check_constraints(name_hash, loaded_class, class_loader, false, THREAD);
-
-        // Need to check for a PENDING_EXCEPTION again; check_constraints
-        // can throw but we may have to remove entry from the placeholder table below.
-        if (!HAS_PENDING_EXCEPTION) {
-          // Record dependency for non-parent delegation.
-          // This recording keeps the defining class loader of the klass (loaded_class) found
-          // from being unloaded while the initiating class loader is loaded
-          // even if the reference to the defining class loader is dropped
-          // before references to the initiating class loader.
-          loader_data->record_dependency(loaded_class);
-
-          { // Grabbing the Compile_lock prevents systemDictionary updates
-            // during compilations.
-            MutexLocker mu(THREAD, Compile_lock);
-            update_dictionary(name_hash, loaded_class, class_loader);
-          }
-
-          if (JvmtiExport::should_post_class_load()) {
-            JvmtiExport::post_class_load(THREAD->as_Java_thread(), loaded_class);
-          }
-        }
-      }
-    } // load_instance_class
+      loaded_class = load_instance_class(name_hash, name, class_loader, THREAD);
```

Comment:
The whole passing THREAD to worry about not returning before removing the placeholder entry is a lot simpler if this code is in its own function, so I moved it there.